libnmstate/nm/context.py

```diff
@@ -62,12 +100,98 @@ def _del_client(self):
                 timeout_source = GLib.timeout_source_new(50)
                 try:
                     timeout_source.set_callback(lambda x: is_timeout.append(1))
-                    timeout_source.attach(self._context)
+                    timeout_source.attach(self.context)
                     while not is_done and not is_timeout:
-                        self._context.iteration(True)
+                        self.context.iteration(True)
                 finally:
                     timeout_source.destroy()
-
             if not is_done:
                 logging.error("BUG: NM.Client is not cleaned")
             self._context = None
+
+    def _del_timeout(self):
+        if self._timeout_source:
+            self._timeout_source.destroy()
+            self._timeout_source = None
+
+    def register_async(self, action, fast=False):
+        """
+        Register action(string) to wait list.
+        """
+        queue = self._fast_queue if fast else self._slow_queue
+        max_queue = FAST_ASYNC_QUEUE_SIZE if fast else SLOW_ASYNC_QUEUE_SIZE
+        if len(queue) >= max_queue:
+            logging.debug(
+                f"Async queue({max_queue}) full, waiting all existing actions "
+                "to be finished before registering more async action"
+            )
+            # TODO: No need to wait all finish, a single finished async is
+            #       sufficient.
+            self.wait_all_finish()
+
+        logging.debug(f"Async action: {action} started")
+        queue.add(action)
+
+    def finish_async(self, action, suppress_log=False):
+        """
+        Mark action(string) as finished.
+        """
+        self._last_async_finish_time = datetime.datetime.now()
+        if not suppress_log:
+            logging.debug(f"Async action: {action} finished")
+        self._fast_queue.discard(action)
+        self._slow_queue.discard(action)
+
+    def _action_all_finished(self):
+        return not (len(self._fast_queue) or len(self._slow_queue))
+
+    def _idle_timeout_cb(self, _user_data):
+        if self._error or self._action_all_finished():
+            return GLib.SOURCE_REMOVE
+        idle_time = datetime.datetime.now() - self._last_async_finish_time
+        if idle_time > datetime.timedelta(IDLE_TIMEOUT):
+            remaining_actions = self._slow_queue | self._fast_queue
+            self.fail(
+                NmstateTimeoutError(f"Action {remaining_actions} timeout")
```

Comment:
It could be nice to add the time elapsed.
But my main question here is: Will we know which action caused this?

Reply:
The time elapsed does not make any sense as all the 100 activation is submitted at the same time, when the timeout happens, the time elapsed is 25 seconds. We don't know which and when NM finish them.

No. We cannot tell, just NM too slow.